### ISSUE TITLE
When the model is in an eval state before calling trainer.fit it should be moved to train state

### ISSUE THREAD

**Issue by Yann-CV** [Lightning-AI/pytorch-lightning#20986]
### Bug description

If somehow the model is set to the `eval` mode (calling `model.eval` or model.train(False)`) before calling teh trainer.fit supposed to trained the model, the model is staying in the `eval` mode, thus not applying some tricks interesting for training.

### What version are you seeing the problem on?

v2.5

### How to reproduce the bug

```python
import torch
from torch.utils.data import Dataset, DataLoader
import pytorch_lightning as pl

# Dummy dataset
class DummyDataset(Dataset):
    def __init__(self, length=5):
        self.length = length
        self.x = torch.randn(length, 10)
        self.y = torch.randn(length, 1)

    def __len__(self):
        return self.length

    def __getitem__(self, idx):
        return self.x[idx], self.y[idx]

# Simple LightningModule
class SimpleModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(10, 1)
        self.loss_fn = torch.nn.MSELoss()

    def forward(self, x):
        return self.linear(x)

    def training_step(self, batch, batch_idx):
        print(f"is the model in train state: {self.training}")
        x, y = batch
        y_hat = self(x)
        loss = self.loss_fn(y_hat, y)
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)

# Instantiate dataset and dataloader
dataset = DummyDataset()
dataloader = DataLoader(dataset, batch_size=16)

# Instantiate model and trainer
model = SimpleModel()
model.eval()
trainer = pl.Trainer(max_epochs=1, logger=False, enable_checkpointing=False)

# Train
trainer.fit(model, dataloader)
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 4060 Ti
                - NVIDIA GeForce GTX 1080 Ti
        - available:         True
        - version:           12.6
* Lightning:
        - lightning:         2.5.0
        - lightning-utilities: 0.14.3
        - pytorch-lightning: 2.5.1.post0
        - torch:             2.7.1
        - torchmetrics:      1.7.3
* Packages:
        - aiohappyeyeballs:  2.6.1
        - aiohttp:           3.12.13
        - aiosignal:         1.3.2
        - alembic:           1.16.2
        - annotated-types:   0.7.0
        - antlr4-python3-runtime: 4.9.3
        - anyio:             4.9.0
        - async-timeout:     5.0.1
        - attrs:             25.3.0
        - autocommand:       2.2.2
        - backports.tarfile: 1.2.0
        - blinker:           1.9.0
        - cachetools:        5.5.2
        - certifi:           2025.6.15
        - charset-normalizer: 3.4.2
        - click:             8.2.1
        - cloudpickle:       3.1.1
        - contourpy:         1.3.2
        - cycler:            0.12.1
        - databricks-sdk:    0.57.0
        - docker:            7.1.0
        - exceptiongroup:    1.3.0
        - fastapi:           0.115.13
        - filelock:          3.18.0
        - flask:             3.1.1
        - fonttools:         4.58.4
        - frozenlist:        1.7.0
        - fsspec:            2025.5.1
        - gitdb:             4.0.12
        - gitpython:         3.1.44
        - google-auth:       2.40.3
        - graphene:          3.4.3
        - graphql-core:      3.2.6
        - graphql-relay:     3.2.0
        - greenlet:          3.2.3
        - gunicorn:          23.0.0
        - h11:               0.16.0
        - idna:              3.10
        - importlib-metadata: 8.7.0
        - inflect:           7.3.1
        - itsdangerous:      2.2.0
        - jaraco.collections: 5.1.0
        - jaraco.context:    5.3.0
        - jaraco.functools:  4.0.1
        - jaraco.text:       3.12.1
        - jinja2:            3.1.6
        - joblib:            1.5.1
        - kiwisolver:        1.4.8
        - lightning:         2.5.0
        - lightning-utilities: 0.14.3
        - mako:              1.3.10
        - markupsafe:        3.0.2
        - matplotlib:        3.10.3
        - mlflow:            3.1.0
        - mlflow-skinny:     3.1.0
        - more-itertools:    10.3.0
        - mpmath:            1.3.0
        - multidict:         6.5.0
        - networkx:          3.4.2
        - numpy:             2.2.6
        - nvidia-cublas-cu12: 12.6.4.1
        - nvidia-cuda-cupti-cu12: 12.6.80
        - nvidia-cuda-nvrtc-cu12: 12.6.77
        - nvidia-cuda-runtime-cu12: 12.6.77
        - nvidia-cudnn-cu12: 9.5.1.17
        - nvidia-cufft-cu12: 11.3.0.4
        - nvidia-cufile-cu12: 1.11.1.6
        - nvidia-curand-cu12: 10.3.7.77
        - nvidia-cusolver-cu12: 11.7.1.2
        - nvidia-cusparse-cu12: 12.5.4.2
        - nvidia-cusparselt-cu12: 0.6.3
        - nvidia-nccl-cu12:  2.26.2
        - nvidia-nvjitlink-cu12: 12.6.85
        - nvidia-nvtx-cu12:  12.6.77
        - omegaconf:         2.3.0
        - opentelemetry-api: 1.34.1
        - opentelemetry-sdk: 1.34.1
        - opentelemetry-semantic-conventions: 0.55b1
        - packaging:         24.2
        - pandas:            2.3.0
        - pillow:            11.2.1
        - platformdirs:      4.2.2
        - propcache:         0.3.2
        - protobuf:          6.31.1
        - pyarrow:           20.0.0
        - pyasn1:            0.6.1
        - pyasn1-modules:    0.4.2
        - pydantic:          2.11.7
        - pydantic-core:     2.33.2
        - pyparsing:         3.2.3
        - python-dateutil:   2.9.0.post0
        - pytorch-lightning: 2.5.1.post0
        - pytz:              2025.2
        - pyyaml:            6.0.2
        - requests:          2.32.4
        - rsa:               4.9.1
        - scikit-learn:      1.7.0
        - scipy:             1.15.3
        - setuptools:        80.9.0
        - six:               1.17.0
        - smmap:             5.0.2
        - sniffio:           1.3.1
        - sqlalchemy:        2.0.41
        - sqlparse:          0.5.3
        - starlette:         0.46.2
        - sympy:             1.14.0
        - tabulate:          0.9.0
        - threadpoolctl:     3.6.0
        - tomli:             2.2.1
        - torch:             2.7.1
        - torchmetrics:      1.7.3
        - tqdm:              4.67.1
        - triton:            3.3.1
        - typeguard:         4.3.0
        - typing-extensions: 4.14.0
        - typing-inspection: 0.4.1
        - tzdata:            2025.2
        - urllib3:           2.5.0
        - uvicorn:           0.34.3
        - werkzeug:          3.1.3
        - wheel:             0.45.1
        - yarl:              1.20.1
        - zipp:              3.23.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.10.14
        - release:           6.8.0-59-generic
        - version:           #61~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 15 17:03:15 UTC 2

</details>

**Comment by bhimrazy** [comment]
From reviewing the [docs](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#validation) and pseudocode, it seems that Lightning assumes the model is already in `train()` mode (which is the default behaviour) when `trainer.fit()` is called, and only toggles back to `train()` after running a `validation` loop. 

```python
# Under the hood, Lightning does the following (pseudocode):
# ...
for batch_idx, batch in enumerate(train_dataloader):
    loss = model.training_step(batch, batch_idx)
    loss.backward()
    # ...

    if validate_at_some_point:
        # disable grads + batchnorm + dropout
        torch.set_grad_enabled(False)
        model.eval()

        # ----------------- VAL LOOP ---------------
        for val_batch_idx, val_batch in enumerate(val_dataloader):
            val_out = model.validation_step(val_batch, val_batch_idx)
        # ----------------- VAL LOOP ---------------

        # enable grads + batchnorm + dropout
        torch.set_grad_enabled(True)
        model.train()
```

Maybe it would be safer if `trainer.fit()` explicitly set `model.train()` at the start of training to avoid this edge case. Curious to hear what @Borda @matsumotosan think  ?

**Comment by deependujha** [comment]
Hi @Yann-CV,

Do you have a specific use case for setting the model to `eval()` before calling `trainer.fit()`?

In general, Lightning assumes you **don’t call** `model.train()` or `model.eval()` directly. 
> It will set the correct mode internally at the right time (e.g., before training, validation, and test loops) and also trigger the relevant hooks such as `on_validation_start`, `on_train_start`, etc.

If your goal is to temporarily disable parameter updates, `model.freeze()` / `model.unfreeze()` is safer than directly calling `model.eval()`.

For finer-grained control over training/eval states, you can also consider **Lightning Fabric**, which gives you more flexibility for custom workflows.

Regards,
Deependu

**Comment by SkafteNicki** [comment]
@bhimrazy @deependujha I found a couple of related issues:
https://github.com/Lightning-AI/pytorch-lightning/issues/20216
https://github.com/Lightning-AI/pytorch-lightning/issues/20573
https://github.com/Lightning-AI/pytorch-lightning/issues/20177

One of the issues specifically mentions the release notes from v2.2 of lightning
https://github.com/Lightning-AI/pytorch-lightning/releases/tag/2.2.0#highlights-eval
where the behavior was changed such that 
> "the Trainer now captures the mode of every submodule before switching to validation, and restores the mode the modules were in when validation ends (https://github.com/Lightning-AI/lightning/pull/18951, https://github.com/Lightning-AI/lightning/pull/18951, https://github.com/Lightning-AI/lightning/pull/18951)

it also have this note

> [!IMPORTANT]
> In some libraries, for example HuggingFace, models are created in evaluation mode by default (e.g. `HFModel.from_pretrained(...)`). Starting from 2.2, you will have to set `.train()` on these models if you intend to train them.

With 4 issues on this topic I think the conclusion is that there is some confusion on this. I would recommend that we:
* Add a warning in `on_fit_start` (or where it is appropriate) that detects if a user provides a model in eval mode. If it is the intended behavior then either ignore warning or overwrite the hook
* Be more specific on this issue in the documentation

cc: @Borda

**Comment by adosar** [comment]
I have opened a PR for the docs: https://github.com/Lightning-AI/pytorch-lightning/pull/20716

cc: @SkafteNicki
